src/relay/transforms/memory_alloc.cc

```diff
@@ -0,0 +1,508 @@
+          CHECK(copy_call);
+          attr = copy_call->attrs;
+        } else {
+          attr = cn->attrs;
+        }
+        const DeviceCopyAttrs* copy_attr = attr.as<DeviceCopyAttrs>();
+        CHECK(copy_attr);
+        return DeviceCopy(new_args[0], copy_attr->src_dev_type, copy_attr->dst_dev_type);
+      } else if (IsDynamic(ret_type)) {
+        Function func = Downcast<Function>(cn->op);
+        return DynamicInvoke(&scope, func, ins, new_args, out_types, ret_type);
+      } else {
+        // Handle the static case
+        Array<Expr> outs;
+        for (size_t i = 0; i < out_types.size(); ++i) {
+          TVMContext ctx = GetContext(GetRef<Call>(cn));
+          auto out = MakeStaticAllocation(&scope, out_types[i], ctx, std::to_string(i));
+          outs.push_back(out);
+        }
+        Tuple output(outs);
+        Expr invoke = (*invoke_tvm_)(cn->op, ins, output);
+        scope.Push(invoke);
+        return ToTupleType(ret_type,
+                           std::vector<Expr>(output->fields.begin(), output->fields.end()));
+      }
+    } else {
+      return ExprMutator::VisitExpr_(cn);
+    }
+  }
+
+ private:
+  // Insert a device copy node.
+  Expr DeviceCopy(const Expr& inp, int src_ctx, int dst_ctx) {
+    return ExprMutator::Mutate((*device_copy_)(inp, src_ctx, dst_ctx));
+  }
+
+  // Check if a call invokes a primitive function.
+  bool IsPrimitive(const CallNode* call) const {
+    if (const auto* fn = call->op.as<FunctionNode>()) {
+      return fn->HasNonzeroAttr(attr::kPrimitive);
+    }
+    return false;
+  }
+
+  // Check if the current relay expression is a device copy call. We can simply
+  // check the body of it if it is a function because the device_copy op is opaque.
+  bool IsDeviceCopy(const Expr& expr) const {
+    if (const auto* fn = expr.as<FunctionNode>()) {
+      auto body = fn->body;
+      const CallNode* call = body.as<CallNode>();
+      return call && call->op == Op::Get("device_copy");
+    } else if (const CallNode* cn = expr.as<CallNode>()) {
+      return cn->op == Op::Get("device_copy");
+    } else {
+      return false;
+    }
+  }
+
+  Expr ComputeAlignment(const DataType& dtype) const {
+    int64_t align = dtype.bits() / 8 * dtype.lanes();
+    if (align < 64) {
+      align = 64;
+    }
+    return MakeConstant(align);
+  }
+
+  Expr ComputeStorageInRelay(const Expr& shape, const TensorType& type) const {
+    auto dtype = DataType(type->dtype);
+    auto fp = runtime::Registry::Get("relay.op._make.prod");
+    CHECK(fp) << "cannot find operator prod from the registry";
+    auto fa = runtime::Registry::Get("relay.op._make.add");
+    CHECK(fa) << "cannot find operator add from the registry";
+    auto fd = runtime::Registry::Get("relay.op._make.divide");
+    CHECK(fd) << "cannot find operator devide from the registry";
+    auto fm = runtime::Registry::Get("relay.op._make.multiply");
+    CHECK(fm) << "cannot find operator multiply from the registry";
```

Comment:
I think it is better to use the existing python function `compute_storage_in_relay` and expose it to c++, rather than doing this way.

Reply:
I dont have a strong preference. Maybe I want to still keep the C++ one. The Python one save a few lines but I don't really want to keep the original file or let the C++ pass call into Python

Reply:
ok, then I think we should `Get` and `CHECK` once in the constructor and cache them, similar to other memory ops.